File: internal/graph/generate.go
package graph

// generateRowSymbols generates the graph symbols for a single commit row.
//
// Parameters:
//   - commitCol: which column this commit is in
//   - numCols: total number of columns to generate
//   - mergeColumns: columns where merge parents are located (for merge lines)
//   - convergingColumns: columns that are converging into this commit (same hash in multiple lanes)
//   - passingColumns: columns with lanes that continue through this row (not commit, not converging)
//   - existingLanesMerge: columns where merge parents existed in lanes (need extra visual column for corner)
//   - convergesToParent: true if this commit converges to its parent (branch ending)
func generateRowSymbols(commitCol int, numCols int, mergeColumns []int, convergingColumns []int, passingColumns []int, existingLanesMerge []int, convergesToParent bool) Row {
	if numCols == 0 {
		return Row{Symbols: []GraphSymbol{}}
	}

	symbols := make([]GraphSymbol, numCols)

	// Initialize all to empty
	for i := range symbols {
		symbols[i] = SymbolEmpty
	}

	// Find rightmost merge or converging column (for drawing merge line)
	rightmostMerge := -1
	for _, col := range mergeColumns {
		if col > rightmostMerge {
			rightmostMerge = col
		}
	}
	rightmostConverge := -1
	for _, col := range convergingColumns {
		if col > rightmostConverge {
			rightmostConverge = col
		}
	}

	// The rightmost column that needs a horizontal line from commit
	rightmostHorizontal := max(rightmostMerge, rightmostConverge)

	// If we have existing lane merges, the corner goes in an extra column beyond
	// the rightmost merge, so extend the horizontal line to that column
	if len(existingLanesMerge) > 0 && rightmostHorizontal >= 0 {
		rightmostHorizontal++
	}

	// If commit converges to parent, the convergence symbol goes in an extra column
	// Note: this is different from existing lane merge - here the line goes LEFT from commit
	// So we need to track where the convergence symbol goes
	convergenceSymbolCol := -1
	if convergesToParent {
		// Convergence symbol goes in the rightmost column (the extra one we added)
		convergenceSymbolCol = numCols - 1
		// Also need horizontal line from commit to convergence
		if commitCol < convergenceSymbolCol {
			rightmostHorizontal = max(rightmostHorizontal, convergenceSymbolCol)
		}
	}

	// Build sets for quick lookup
	mergeSet := make(map[int]bool)
	for _, col := range mergeColumns {
		mergeSet[col] = true
	}
	convergeSet := make(map[int]bool)
	for _, col := range convergingColumns {
		convergeSet[col] = true
	}
	passingSet := make(map[int]bool)
	for _, col := range passingColumns {
		passingSet[col] = true
	}
	existingLaneMergeSet := make(map[int]bool)
	for _, col := range existingLanesMerge {
		existingLaneMergeSet[col] = true
	}

	// Generate symbols for each column
	for col := 0; col < numCols; col++ {
		if col == commitCol {
			// This is the commit column
			if convergesToParent {
				// Commit converges to parent - show convergence line
				symbols[col] = SymbolMergeCommit // ├─ (will show ├─╯ when combined with convergence symbol)
			} else if rightmostHorizontal > commitCol {
				// There's a merge line going right
				symbols[col] = SymbolMergeCommit // ├─
			} else {
				symbols[col] = SymbolCommit // ├
			}
		} else if mergeSet[col] && convergeSet[col] && !existingLaneMergeSet[col] {
			// Column is BOTH merging AND converging - merge join
			// This happens when a merge parent reuses a converging column
			if col < rightmostHorizontal {
				symbols[col] = SymbolMergeCross // ┼─ (continues right)
			} else {
				symbols[col] = SymbolMergeJoin // ┤ (rightmost)
			}
		} else if mergeSet[col] && !existingLaneMergeSet[col] {
			// This is a merge parent column (new branch starting from merge)
			if col < rightmostMerge {
				symbols[col] = SymbolOctopus // ┬─ (continues right to more merges)
			} else {
				symbols[col] = SymbolBranchTop // ╮ (rightmost merge)
			}
		} else if convergeSet[col] {
			// This is a converging column (branch ending, joining commit)
			if col < rightmostConverge {
				symbols[col] = SymbolDiverge // ┴─ (continues right to more convergences)
			} else {
				symbols[col] = SymbolBranchBottom // ╯ (rightmost convergence)
			}
		} else if passingSet[col] || existingLaneMergeSet[col] {
			// This is a passing lane (or an existing lane being merged)
			if col > commitCol && col < rightmostHorizontal {
				// Merge line crosses this vertical line
				symbols[col] = SymbolBranchCross // │─
			} else {
				symbols[col] = SymbolBranchPass // │
			}
		} else if col > commitCol && col < rightmostHorizontal {
			// Empty column in merge line path - could be the extra corner column
			symbols[col] = SymbolEmpty
		}

		// Special case: if this is the rightmost column and we have existing lane merges,
		// it might be the extra corner column
		if col == rightmostHorizontal && len(existingLanesMerge) > 0 {
			symbols[col] = SymbolBranchTop // ╮
		}

		// Special case: if this is the convergence symbol column
		if col == convergenceSymbolCol {
			symbols[col] = SymbolBranchBottom // ╯
		}
	}

	return Row{Symbols: symbols}
}

// max returns the larger of two integers.
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// detectConvergingColumns finds columns that have the same hash as the commit.
// This indicates branches that are converging at this commit (common ancestor).
//
// Parameters:
//   - commitCol: the column where the commit is located
//   - commitHash: the hash of the current commit
//   - lanes: current lane state (before this commit is processed)
//
// Returns indices of lanes (other than commitCol) that have commitHash.
func detectConvergingColumns(commitCol int, commitHash string, lanes []string) []int {
	var converging []int
	for i, hash := range lanes {
		if i != commitCol && hash == commitHash {
			converging = append(converging, i)
		}
	}
	return converging
}

// detectPassingColumns finds columns that have active lanes passing through
// (not the commit column, not merging, not converging).
//
// Parameters:
//   - commitCol: the column where the commit is located
//   - lanes: current lane state
//   - mergeColumns: columns that are merge parents
//   - convergingColumns: columns that are converging
//
// Returns indices of lanes that are just passing through.
func detectPassingColumns(commitCol int, lanes []string, mergeColumns []int, convergingColumns []int) []int {
	// Build exclusion set
	exclude := make(map[int]bool)
	exclude[commitCol] = true
	for _, col := range mergeColumns {
		exclude[col] = true
	}
	for _, col := range convergingColumns {
		exclude[col] = true
	}

	var passing []int
	for i, hash := range lanes {
		if !exclude[i] && hash != "" {
			passing = append(passing, i)
		}
	}
	return passing
}

File: src/domain/graph/generate.rs
use super::types::{GraphSymbol, Row};

pub fn generate_row_symbols(
    commit_col: usize,
    num_cols: usize,
    merge_columns: &[usize],
    converging_columns: &[usize],
    passing_columns: &[usize],
    existing_lanes_merge: &[usize],
    converges_to_parent: bool,
) -> Row {
    if num_cols == 0 {
        return Row { symbols: vec![] };
    }

    let mut symbols = vec![GraphSymbol::Empty; num_cols];

    let rightmost_merge = merge_columns.iter().copied().max();
    let rightmost_converge = converging_columns.iter().copied().max();

    let mut rightmost_horizontal = match (rightmost_merge, rightmost_converge) {
        (Some(m), Some(c)) => std::cmp::max(m, c),
        (Some(m), None) => m,
        (None, Some(c)) => c,
        (None, None) => commit_col,
    };

    if !existing_lanes_merge.is_empty() {
        rightmost_horizontal += 1;
    }

    let convergence_symbol_col = if converges_to_parent {
        Some(num_cols - 1)
    } else {
        None
    };

    if let Some(conv_col) = convergence_symbol_col {
        if commit_col < conv_col {
            rightmost_horizontal = std::cmp::max(rightmost_horizontal, conv_col);
        }
    }

    let merge_set: std::collections::HashSet<usize> = merge_columns.iter().copied().collect();
    let converge_set: std::collections::HashSet<usize> =
        converging_columns.iter().copied().collect();
    let passing_set: std::collections::HashSet<usize> = passing_columns.iter().copied().collect();
    let existing_lane_merge_set: std::collections::HashSet<usize> =
        existing_lanes_merge.iter().copied().collect();

    for (col, symbol) in symbols.iter_mut().enumerate() {
        if col == commit_col {
            if converges_to_parent || rightmost_horizontal > commit_col {
                *symbol = GraphSymbol::MergeCommit;
            } else {
                *symbol = GraphSymbol::Commit;
            }
        } else if merge_set.contains(&col)
            && converge_set.contains(&col)
            && !existing_lane_merge_set.contains(&col)
        {
            if col < rightmost_horizontal {
                *symbol = GraphSymbol::MergeCross;
            } else {
                *symbol = GraphSymbol::MergeJoin;
            }
        } else if merge_set.contains(&col) && !existing_lane_merge_set.contains(&col) {
            if let Some(rm) = rightmost_merge {
                if col < rm {
                    *symbol = GraphSymbol::Octopus;
                } else {
                    *symbol = GraphSymbol::BranchTop;
                }
            } else {
                *symbol = GraphSymbol::BranchTop;
            }
        } else if converge_set.contains(&col) {
            if let Some(rc) = rightmost_converge {
                if col < rc {
                    *symbol = GraphSymbol::Diverge;
                } else {
                    *symbol = GraphSymbol::BranchBottom;
                }
            } else {
                *symbol = GraphSymbol::BranchBottom;
            }
        } else if passing_set.contains(&col) || existing_lane_merge_set.contains(&col) {
            if col > commit_col && col < rightmost_horizontal {
                *symbol = GraphSymbol::BranchCross;
            } else {
                *symbol = GraphSymbol::BranchPass;
            }
        }

        if col == rightmost_horizontal && !existing_lanes_merge.is_empty() {
            *symbol = GraphSymbol::BranchTop;
        }

        if Some(col) == convergence_symbol_col {
            *symbol = GraphSymbol::BranchBottom;
        }
    }

    Row { symbols }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_row_commit_only() {
        let row = generate_row_symbols(0, 1, &[], &[], &[], &[], false);
        assert_eq!(row.symbols, vec![GraphSymbol::Commit]);
    }

    #[test]
    fn test_generate_row_with_merge() {
        let row = generate_row_symbols(0, 2, &[1], &[], &[], &[], false);
        assert_eq!(
            row.symbols,
            vec![GraphSymbol::MergeCommit, GraphSymbol::BranchTop]
        );
    }

    #[test]
    fn test_generate_row_with_convergence() {
        let row = generate_row_symbols(0, 2, &[], &[1], &[], &[], false);
        assert_eq!(
            row.symbols,
            vec![GraphSymbol::MergeCommit, GraphSymbol::BranchBottom]
        );
    }
}
